rust/arrow/src/array/array.rs

```diff
@@ -409,4 +578,42 @@ mod tests {
         assert_eq!(a.len(), 0);
         assert_eq!(a.value_offsets()[0], 0i32);
     }
+
+    #[test]
+    fn test_null_boolean() {
+        let array = new_array_with_nulls(&DataType::Boolean, 9);
+        let a = array.as_any().downcast_ref::<BooleanArray>().unwrap();
+        assert_eq!(a.len(), 9);
+        for i in 0..9 {
+            assert!(a.is_null(i));
+        }
+    }
+
+    #[test]
+    fn test_null_primitive() {
+        let array = new_array_with_nulls(&DataType::Int32, 9);
+        let a = array.as_any().downcast_ref::<Int32Array>().unwrap();
+        assert_eq!(a.len(), 9);
+        for i in 0..9 {
+            assert!(a.is_null(i));
+        }
+    }
+
+    #[test]
+    fn test_null_variable_sized() {
+        let array = new_array_with_nulls(&DataType::Utf8, 9);
+        let a = array.as_any().downcast_ref::<StringArray>().unwrap();
+        assert_eq!(a.len(), 9);
+        assert_eq!(a.value_offsets()[9], 0i32);
+    }
+
+    #[test]
+    fn test_null_list_primitive() {
+        let data_type =
+            DataType::List(Box::new(Field::new("item", DataType::Int32, true)));
```

Comment:
There's a soundness hole here, that applies more for parquet. If I created this array as non-nullable, there wouldn't have been an error, even though this doesn't comply with the spec.

@alamb @jorgecarleitao @dandandan should we panic, or should I change the signature to `Result<ArrayRef>` so I can validate this condition?

I'm leaning to a panic, just like we panic on `impl From<ArrayDataRef> for {type}Array`. Wrapping a function that doesn't really error out, in a Result feels like assembing the Avengers for a single villain.

Reply:
I vote `panic` for now if you think it is a logic error that a user should be avoiding anyways.

Or alternately, we can make it `panic` now and `Result` later if someone finds a need

Reply:
yeah, that hole exists in so many places, unfortunately. Since `ArrayData` has a `null_count`, we have been relying on it for this.

Technically, this goes back to the issue that we are not performing any validation whether the datatype is consistent with the array's content. :(